### PR TITLE
Rewrite aes encryption #257

### PR DESCRIPTION
**Pull request by kangpinghuang**

### CODE REVIEW

File: be/src/aes/aes.cpp [hunk 1]
```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2018, Baidu.com, Inc. All Rights Reserved
```
Comment (imay) [inline be/src/aes/aes.cpp]: please use ASF's header

File: be/CMakeLists.txt [hunk 2]
```diff
@@ -585,6 +585,7 @@ if (${MAKE_TEST} STREQUAL "ON")
     add_subdirectory(${TEST_DIR}/exprs)
     add_subdirectory(${TEST_DIR}/runtime)
     add_subdirectory(${TEST_DIR}/http)
+    add_subdirectory(${TEST_DIR}/aes)
```
Comment (imay) [inline be/CMakeLists.txt]: I think we don't need this 'aes' directory

you can put aes.h and aes.cpp to be/src/util, and rename these to aes_util.h??

File: be/src/aes/aes.cpp [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    EVP_CIPHER_CTX_cleanup(&aes_ctx);
+    if (!ret) {
+        ERR_clear_error();
+        return AES_BAD_DATA;
+    } else {
+        return u_len + f_len;
+    }
+}
+
+int do_decrypt(EVP_CIPHER_CTX* aes_ctx, const EVP_CIPHER* cipher,
```
Comment (imay) [inline be/src/aes/aes.cpp]: this function should be static

**Review by imay — CHANGES_REQUESTED** [review]

File: be/src/util/aes_util.cpp [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        const unsigned char* key, uint32_t key_length, const unsigned char* iv,
+        bool padding, unsigned char* encrypt) {
+    EVP_CIPHER_CTX aes_ctx;
+    const EVP_CIPHER* cipher= get_evp_type(mode);
+    int u_len, f_len;
+    /* The encrypt key to be used for encryption */
+    unsigned char encrypt_key[AES_MAX_KEY_LENGTH / 8];
+    aes_create_key(key, key_length, encrypt_key, mode);
+    
+    if (!cipher || (EVP_CIPHER_iv_length(cipher) > 0 && !iv)) {
```
Comment (imay) [inline be/src/util/aes_util.cpp]: when compare pointer use "cipher != nullptr", compare integer use "ret != 0", check bool use "!is_success"

File: be/src/util/aes_util.cpp [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        return AES_BAD_DATA;
```
Comment (imay) [inline be/src/util/aes_util.cpp]: when cipher is not null, cipher's memory leak

File: be/src/util/aes_util.cpp [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+#include "exprs/base64.h"
+
+namespace palo {
+
+static const int AES_MAX_KEY_LENGTH = 256;
+
+const EVP_CIPHER* get_evp_type(const AesMode mode) {
+    switch (mode) {
+        case AES_128_ECB:
```
Comment (imay) [inline be/src/util/aes_util.cpp]: the indent is not right, 'case" align with switch, no need to indent

**Review by imay — CHANGES_REQUESTED** [review]

File: be/src/util/aes_util.cpp (hunk 5 above)
Comment (kangpinghuang) [reply be/src/util/aes_util.cpp]: cipher can not be delete by manually or it will core dump

**Review by imay — DISMISSED** [review]
LGTM

**Review by imay — APPROVED** [review]

**Pull request merged**